src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java

```diff
@@ -0,0 +1,105 @@
+import com.atlassian.bitbucket.jenkins.internal.model.*;
+import com.fasterxml.jackson.core.type.TypeReference;
+import hudson.model.TaskListener;
+import okhttp3.HttpUrl;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+import static com.atlassian.bitbucket.jenkins.internal.util.SystemPropertiesConstants.REMOTE_TAGS_RETRIEVAL_MAX_PAGES;
+import static com.atlassian.bitbucket.jenkins.internal.util.SystemPropertiesConstants.REMOTE_TAGS_RETRIEVAL_PAGE_SIZE;
+import static java.lang.String.valueOf;
+
+public class BitbucketTagClientImpl implements BitbucketTagClient {
+
+    private static final int MAX_PAGES = Integer.getInteger(REMOTE_TAGS_RETRIEVAL_MAX_PAGES, 5);
+    private static final int PAGE_SIZE = Integer.getInteger(REMOTE_TAGS_RETRIEVAL_PAGE_SIZE, 1000);
+
+    private final BitbucketRequestExecutor bitbucketRequestExecutor;
+    private final String projectKey;
+    private final String repositorySlug;
+    private final TaskListener taskListener;
+
+    public BitbucketTagClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor,
+                                  String projectKey,
+                                  String repositorySlug,
+                                  TaskListener taskListener) {
+        this.bitbucketRequestExecutor = bitbucketRequestExecutor;
+        this.projectKey = projectKey;
+        this.repositorySlug = repositorySlug;
+        this.taskListener = taskListener;
+    }
+
+    @Override
+    public Stream<BitbucketTag> getRemoteTags() {
+        HttpUrl.Builder urlBuilder = bitbucketRequestExecutor.getCoreRestPath().newBuilder()
+                .addPathSegment("projects")
+                .addPathSegment(projectKey)
+                .addPathSegment("repos")
+                .addPathSegment(repositorySlug)
+                .addPathSegment("tags")
+                .addQueryParameter("limit", String.valueOf(PAGE_SIZE))
+                .addQueryParameter("orderBy", "modification"); // the most recently-modified tags first
+
+        HttpUrl url = urlBuilder.build();
+        BitbucketPage<BitbucketTag> firstPage =
+                bitbucketRequestExecutor.makeGetRequest(url,
+                        new TypeReference<BitbucketPage<BitbucketTag>>() {
+                        }).getBody();
+
+        return BitbucketPageStreamUtil.toStream(firstPage, new BitbucketTagClientImpl.NextPageFetcherImpl(url, bitbucketRequestExecutor,
+                        MAX_PAGES, taskListener))
+                .map(BitbucketPage::getValues)
+                .flatMap(Collection::stream);
+    }
+
+    static class NextPageFetcherImpl implements NextPageFetcher<BitbucketTag> {
+
+        private final BitbucketRequestExecutor bitbucketRequestExecutor;
+        private final AtomicInteger currentPage = new AtomicInteger();
+        private final int maxPages;
```

Comment:
I don't think you need this- `NextPageFetcherImpl` has private access to it's containing class so you should be able to reference `MAX_PAGES`

Reply:
We do it this way in branchClient and PRClient and are warned that we need a synthetic accessor.

Reply:
Is there any reason a synthetic accessor is a concern here?

Reply:
I think it was mostly done to make testing unit testing easier by being able to use different mock values.